### PR TITLE
Refactor TTIRLayoutOperandsRewriter pass

### PR DESCRIPTION
**Pull request by nsmithtt**
Change the pass to pattern match on DPS interface instead, it's a generic rewrite for all DPS ops anyway.

Default the input and output graph arguments to L1, the runtime will automatically upload them there.

### CODE REVIEW

**Comment by nsmithtt** [comment]
Most of the tests had to change because of the defaulting to L1.  This no longer requires a `to_layout`/`to_memory_config` because the inputs are allowed to start in L1.  Runtime has support for automatically moving inputs there.

File: test/ttmlir/Dialect/TTNN/ttir_to_ttnn_pipeline.mlir [hunk 1]
```diff
@@ -2,13 +2,12 @@
 #any_device = #tt.operand_constraint<dram|l1|scalar|tile|any_device|any_device_tile>
 module attributes {} {
   func.func @forward(%arg0: tensor<64x128xf32>, %arg1: tensor<64x128xf32>) -> tensor<64x128xf32> {
-    // CHECK: #layout2 = #tt.layout<(d0, d1) -> (d0, d1), undef, <8x8>, memref<8x16xf32, #l1_>>
+    // CHECK: #[[LATOUT_1:.*]] = #tt.layout<(d0, d1) -> (d0, d1), undef, <8x8>, memref<8x16xf32, #l1_>>
```
Comment (nobradovictt) [inline test/ttmlir/Dialect/TTNN/ttir_to_ttnn_pipeline.mlir:5]: letout :)

**Comment by nobradovictt** [comment]
> Change the pass to pattern match on DPS interface instead, it's a generic rewrite for all DPS ops anyway.
> 
> Default the input and output graph arguments to L1, the runtime will automatically upload them there.

Could you clarify a bit. What happens if there is no space in L1? Will it auto fallback to DRAM interleaved or fail?

**Comment by nvukobratTT** [comment]
> > Change the pass to pattern match on DPS interface instead, it's a generic rewrite for all DPS ops anyway.
> > Default the input and output graph arguments to L1, the runtime will automatically upload them there.
> 
> Could you clarify a bit. What happens if there is no space in L1? Will it auto back to DRAM interleaved or fail?

If we're defaulting to DRAM if there isn't enough memory, it'll be nice to report that as a warning/debug log to the user (perf impact to the model). 

In other cases, if we're erroring out, it'll be nice to provide more details to the end user, with recommended solutions.

**Comment by nsmithtt** [comment]
> > Change the pass to pattern match on DPS interface instead, it's a generic rewrite for all DPS ops anyway.
> > Default the input and output graph arguments to L1, the runtime will automatically upload them there.
> 
> Could you clarify a bit. What happens if there is no space in L1? Will it auto fallback to DRAM interleaved or fail?

This is a problem for optimizer :) currently we don't have any algorithm for spilling tensors.  But in the early stages of the graph we need layout and therefore we need to pick some memory space to start.  I think we should make everything L1 to start (not care about L1 size) just during the graph transformation part.  Then it will be handed over to optimizer which has to decide which memory space they can live in, decide which ones to spill and actually allocate addresses for them in both DRAM and L1.

**Comment by nsmithtt** [comment]
Addressing both comments here regarding memory.  This change is not attempting to choose memory space where tensors can legally fit or not, it's just changing a default.  Previously the inputs were defaulted to system address memory space and had to use to_layout to move them into L1, so we were just as likely to OOM.  This is leveraging the runtime work that was done to automatically move tensors into L1.

**Comment by nobradovictt** [comment]
> Addressing both comments here regarding memory. This change is not attempting to choose memory space where tensors can legally fit or not, it's just changing a default. Previously the inputs were defaulted to system address memory space and had to use to_layout to move them into L1, so we were just as likely to OOM. This is leveraging the runtime work that was done to automatically move tensors into L1.
> 
> Choosing a legal solution is much more involved and will require a proper allocator and algorithm in the optimizer.

Ok so here is the main question, obviously not tied directly to your PR here since I agree situation stays the same, why are we not choosing safest option as default and thats DRAM interleaved(thats how everyone begins model bringup anyway)?  
I mean we will fix this in optimizer but in the meantime some model experiments could fail. Also if Optimizer has some issues and you want to turn it off to test something else, you will still end up with broken compiler. I mean maybe I am pessimistically overoptimizing :) here but this may cause some friction in the early development phase of this project.
Anyway as I said not a blocker for this PR.

**Review by nobradovictt — APPROVED** [review]

**Review by nvukobratTT — APPROVED** [review]

**Comment by nsmithtt** [comment]
> Ok so here is the main question, obviously not tied directly to your PR here since I agree situation stays the same, why are we not choosing safest option as default and thats DRAM interleaved(thats how everyone begins model bringup anyway)?
> I mean we will fix this in optimizer but in the meantime some model experiments could fail. Also if Optimizer has some issues and you want to turn it off to test something else, you will still end up with broken compiler. I mean maybe I am pessimistically overoptimizing :) here but this may cause some friction in the early development phase of this project.
> Anyway as I said not a blocker for this PR.

I see your point, probably most cases would just work out of the box will dram interleaved.  There are currently 2 issues:
- Direct to metal path only supports L1, I'm kind of overfitting here, maybe I can change the PR to allow the backend to program which default memory space the tensor should live.
- We don't have a way to specify interleaved at the moment.  Direct to metal path will never use interleaved layout, it will always abide by affine map, you can program an affine map to have interleaved access if you want.  It's also unclear to me if interleaved is ever desirable, so I didn't want to invest too much effort in modeling this in compiler.  I was hoping to only support sharded, but perhaps some ops only support interleaved? Not sure.

**Pull request merged**